docs/blog/2024-11-06-chicory-1.0.0-M1.md

```diff
@@ -0,0 +1,319 @@
+
+This feature requires you to depend on the experimental annotation processor:
+
+```xml
+<dependency>
+  <groupId>com.dylibso.chicory</groupId>
+  <artifactId>host-module-annotations-experimental</artifactId>
+</dependency>
+
+You can now annotate a class as `@HostModule("your-wasm-module-name")` 
+and export instance members of such a class by annotating them as `@WasmExport`.
+
+They will generate a `HostFunction` called `wasm_module_name.your_function_name`.
+
+For instance, the following generates a HostFunction called wasi_snapshot_preview1.fd_close:
+
+```java
+@HostModule("wasi_snapshot_preview1")
+public final class WasiPreview1 {
+    ...
+    @WasmExport
+    public int fdClose(int fd) {
+        ...
+    }
+}
+```
+
+As you might have noticed from the example above, we are now using this feature ourselves
+to implement the [WASI host functions][wasi].
+
+The annotation processor generates a companion `<YourClassName>_ModuleFactory`, with a
+static member `<YourClassName>_ModuleFactory#toHostFunctions(<YourClassName>)`.
+
+This static member returns an array of `HostFunction`s that perform the
+conversion between raw longs to the type signature you have declared.
+
+You can expose this member in the most convenient way for your end users.
+
+For instance in this case:
+
+```java
+@HostModule("wasi_snapshot_preview1")
+public final class WasiPreview1 {
+    ...
+
+    public HostFunction[] toHostFunctions() {
+        return WasiPreview1_ModuleFactory.toHostFunctions(this);
+    }
+    ...
+}
+```
+
+### Supported Types
+
+Supported input and output types for `@WasmExport` are primitives `int`, `long`, `float`, `double`,
+which correspond to WASM types `I32`, `I64`, `F32`, `F64`.
+
+We also support `java.lang.String` arguments, provided that they are only passed as an input value,
+and they are annotated. We support two annotations:
+
+- `@Buffer`: the code generator translates the `String` value to a pair of integers:
+  - the first being the index of (or pointer to) the beginning of the string UTF-8 bytes in the linear memory
+  - the second being the length of the string in bytes
+- `@CString`: the code generator translates the `String` into an index (pointer to) the beginning 
+   of a null-terminated string of UTF-8 bytes
+
+For instance:
+
+```java
+    @WasmExport
+    public int pathFilestatSetTimes(
+            int fd,
+            int lookupFlags,
+            @Buffer String rawPath,
+            long accessTime,
+            long modifiedTime,
+            int fstFlags) {
+        ...
+    }
+```
+
+is effectively translated to the host function `wasi_snapshot_preview1.path_filestat_set_times` with signature:
+- `long fd            // ValueType.I32`
+- `long lookup_flags  // ValueType.I32`
+- **`long raw_path_ptr  // ValueType.I32`**
+- **`long raw_path_size // ValueType.I32`**
+- `long access_time   // ValueType.I64`
+- `long modified_time // ValueType.I64`
+- `long fst_flags     // ValueType.I32`
+- returns `long[]`
```

Comment:
Let's see if this is more readable:
```java
public long[] path_filestat_set_times(
  long fd            // ValueType.I32
  long lookup_flags  // ValueType.I32
  long raw_path_ptr  // ValueType.I32 => string pointer
  long raw_path_size // ValueType.I32 => string length
  long access_time   // ValueType.I64
  long modified_time // ValueType.I64
  long fst_flags     // ValueType.I32
)

Reply:
nice, but then I am rewording the sentence above:

> with a signature equivalent to

the reason it was not "code-structured" is that technically the signature would be still `(Interface, long...)` :^)

`````
is translated to the host function `wasi_snapshot_preview1.path_filestat_set_times` 
with a signature equivalent to:

```
public long[] path_filestat_set_times(
    long fd            // ValueType.I32
    long lookup_flags  // ValueType.I32
    long raw_path_ptr  // ValueType.I32 => string pointer
    long raw_path_size // ValueType.I32 => string length
    long access_time   // ValueType.I64
    long modified_time // ValueType.I64
    long fst_flags     // ValueType.I32
)
```

`````